### PR TITLE
70 change main inference model to multi label version

### LINKED ISSUE #70: Change main inference model to multi-label version
### Overview

We have developed a prompt & model that generates 5 labels for each claim:
understandability, type_of_claim , type_of_medical_claim , support , harm

We should switch the model used in the MVP to use this version instead, as a precursor to fixing #61 as it will allow us to display and sort by the degree of checkworthiness.

We potentially have the option of using a fine-tuned Gemini model here, or using in-context learning (i.e. putting the training data in the prompt).  The results should be broadly similar so let's deploy whatever is simplest to deploy!

### Requirements

- [ ] switch the model in `vertex.py/generate_reponse()` to the new model
- [ ] make sure the output (a JSON object) is stored in the `inferred_claims` table - the output will need to be converted from JSON to a single string first

### PR DESCRIPTION
**Pull request by dcorney**
Fixes #70 .

Switches over from an off-the-shelf Gemini model to using one trained on our own annotated data set.
This new model also produces much richer labels as output.

Should probably be merged after the database changes of #71 .

-----------------

## Pull request checklist

- [x] I have linked my PR to an issue
- [x] I’ve used [conventional commits](https://github.com/FullFact/automation-docs/wiki/Conventional-commits)
- [x] My branch is up-to-date with `main`
- [ ] Where appropriate, I have added or updated tests
- [ ] Where appropriate, I have [updated documentation](https://github.com/FullFact/automation-docs/) to reflect my changes

### CODE REVIEW

**Review by andylolz — CHANGES_REQUESTED** [review]
Sorry sorry, I realise I am once again reviewing a draft PR. Just flagging a couple of things.

File: src/raphael_backend_flask/app.py [hunk 1]
```diff
@@ -134,19 +137,27 @@ def post_transcripts() -> ResponseReturnValue:
...
-        )
+    inferred_claims = infer_claims(video_id, transcript)
+    # returns list of dicts, each being {'response':[{"claim":... "labels:{...}"}], "chunk"}
+
+    for response in inferred_claims:
+        claims = response.get("response")
+        chunk = response.get("chunk")
+        for claim in claims:
+            label_str = json.dumps(claim.get("labels", {}))
+            checkworthiness = claim.get("labels", {}).get("summary", "na")
```
[the remark this answers is not in the document]
Comment (dcorney) [reply src/raphael_backend_flask/app.py:149]: I'll leave it for now: it's not currently used, but will at some point be needed in the the front end. This line stands as a comment showing how to access the relevant field.

File: src/raphael_backend_flask/app.py [hunk 2]
```diff
@@ -134,19 +137,27 @@ def post_transcripts() -> ResponseReturnValue:
...
+            # checkworthiness will be one of "worth checking", "may be worth checking" or "not worth checking"
+            execute_sql(
+                "INSERT INTO inferred_claims (video_id, claim, label, model, offset_ms) VALUES (?, ?, ?, ?, ?)",
+                (
+                    video_id,
+                    claim["claim"],
+                    label_str,
+                    "gemini-pro",
+                    chunk["start_offset"],
```
[the remark this answers is not in the document]
Comment (dcorney) [reply src/raphael_backend_flask/app.py:158]: Fair point, will fix.

File: src/raphael_backend_flask/app.py [hunk 3]
```diff
@@ -134,19 +137,27 @@ def post_transcripts() -> ResponseReturnValue:
...
+            label_str = json.dumps(claim.get("labels", {}))
+            checkworthiness = claim.get("labels", {}).get("summary", "na")
+            # checkworthiness will be one of "worth checking", "may be worth checking" or "not worth checking"
+            execute_sql(
+                "INSERT INTO inferred_claims (video_id, claim, label, model, offset_ms) VALUES (?, ?, ?, ?, ?)",
+                (
+                    video_id,
+                    claim["claim"],
+                    label_str,
+                    "gemini-pro",
```
[the remark this answers is not in the document]
Comment (dcorney) [reply src/raphael_backend_flask/app.py:157]: Might raise a separate issue on this so I don't forget.

[the remark this answers is not in the document]
Comment (andylolz) [reply src/raphael_backend_flask/app.py:156]: +1, separate issue is fine/good.

File: src/raphael_backend_flask/app.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (andylolz) [reply src/raphael_backend_flask/app.py:148]: Sounds good

**Comment by andylolz** [comment]
I’ve changed the base branch here so that it points at `dev` instead, and I’ve brought it up-to-date with the work on `dev`.

File: src/health_misinfo_shared/fine_tuning.py [hunk 4]
```diff
@@ -425,6 +427,27 @@ def construct_in_context_examples(data_filenames: list[str]) -> tuple[str, list[
     return examples, hold_out_set
 
 
+def infer_claims(video_id: str, transcript: list[dict]) -> Generator:
+    """For use in app"""
```
Comment (andylolz) [inline src/health_misinfo_shared/fine_tuning.py:431]: To get this to work locally, I need to add the following here:
```python
vertexai.init(project=GCP_PROJECT_ID, location=GCP_TUNED_MODEL_LOCATION)
```

Is that something that’s missing here? Or is this a problem of my local setup?

Comment (andylolz) [reply src/health_misinfo_shared/fine_tuning.py:431]: I presume I must have a default project set up on my system somewhere, which differs from `GCP_PROJECT_ID`.

For this to work in production, though, I suspect we do need this extra line.

Comment (ff-dh) [reply src/health_misinfo_shared/fine_tuning.py:431]: I'm going to tentatively assume project should be pulled from the metadata server and so is a you problem, while location might be something we want to add as an argument.

File: src/health_misinfo_shared/fine_tuning.py [hunk 5]
```diff
@@ -264,7 +264,7 @@ def get_model_by_display_name(display_name: str) -> TextGenerationModel:
 
 def get_video_responses(
     model, chunks: list[str], multilabel: bool = False, in_context_examples: str = ""
-) -> None:
+) -> Generator:
```
Comment (ff-dh) [inline src/health_misinfo_shared/fine_tuning.py:267]: I believe this could be a well-typed as Iterable (`Iterable[dict]` at least), bit clearer than a Generator IMO

File: src/health_misinfo_shared/fine_tuning.py (hunk 4 above)
Comment (andylolz) [reply src/health_misinfo_shared/fine_tuning.py:431]: Aha, okay – so I think if I run the `gcloud` CLI I can see where this project is currently specified:

```shell
$ gcloud config list

[core]
account = andy.lulham@fullfact.org
disable_usage_reporting = True
project = fullfact-development-2
```

I guess I can override this by adding `GOOGLE_APPLICATION_CREDENTIALS` to my local .env file.

There ought to be something in the README about this, because it’s not obvious currently.

**Review by ff-dh — COMMENTED** [review]
I think Andy's comment is accurate, that we should get the fine tuned model location in here (either in code, or if you want to the env).

File: src/health_misinfo_shared/fine_tuning.py (hunk 4 above)
Comment (ff-dh) [reply src/health_misinfo_shared/fine_tuning.py:431]: I don't know if this project's readme is the right place because this happens in any project where we have a Google cloud library using a GCP project ID, just usually that service is available on production and dev in all regions - but you are right it should be written down somewhere

File: src/health_misinfo_shared/fine_tuning.py (hunk 5 above)
Comment (andylolz) [reply src/health_misinfo_shared/fine_tuning.py:267]: Yeah good point, `Generator` was a bit lazy on my part!

Comment (andylolz) [reply src/health_misinfo_shared/fine_tuning.py:267]: I think this could be: `Generator[dict[str, Any], None, None]`

Does that sound okay?

Comment (ff-dh) [reply src/health_misinfo_shared/fine_tuning.py:267]: I don't see the aversion to Iterable (Generator subclasses Iterable iirc) but that too sounds okay

Comment (andylolz) [reply src/health_misinfo_shared/fine_tuning.py:267]: > Generator subclasses Iterable iirc

^^ This is the reason I’ve preferred Generator here. I.e. the typing should be as specific as possible. We’re not actually using any of the extra features of a generator, so iterable is fine here.

Comment (andylolz) [reply src/health_misinfo_shared/fine_tuning.py:267]: Okay – I’ve updated all instances of `Generator` with `Iterable`

File: src/health_misinfo_shared/fine_tuning.py (hunk 4 above)
Comment (andylolz) [reply src/health_misinfo_shared/fine_tuning.py:431]: My immediate preference would be to add this line, because it’s a quick fix and it’s what we do elsewhere in this project already anyway, e.g.:
https://github.com/FullFact/health-misinfo-shared/blob/840e2a939f82b65e1519ed1644436098239b1c8f/src/health_misinfo_shared/vertex.py#L28

But I definitely agree with Dakota that hardcoding the project ID like this is not ideal. I think this ought to be set in the environment.

Comment (andylolz) [reply src/health_misinfo_shared/fine_tuning.py:431]: Discussed with Dakota and we’re in agreement (that this is okay for MVP but not ideal) so I’ve pushed this change.

Comment (dcorney) [reply src/health_misinfo_shared/fine_tuning.py:431]: For the MVP, I'm happy to hard code these values as they won't/can't change. But definitely one to plan more robustly when we build the actual prodct.

File: src/raphael_backend_flask/app.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (andylolz) [reply src/raphael_backend_flask/app.py]: Ah – I’ve just realised this is what #89 is about! Great – thank you

**Review by ff-dh — APPROVED** [review]

**Review by andylolz — APPROVED** [review]
Approving this, but probably good for Dakota/David to review as I’ve pushed to this branch most recently.

File: src/raphael_backend_flask/app.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (dcorney) [reply src/raphael_backend_flask/app.py]: Yes! Thank you for adding the link to 89; should have done that myself.
